File: app/src/main/java/com/smoothradio/radio/core/di/AppModule.java
package com.smoothradio.radio.core.di;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.room.Room;

import com.google.firebase.firestore.FirebaseFirestore;
import com.smoothradio.radio.core.data.local.AppDatabase;
import com.smoothradio.radio.core.data.local.RadioStationDao;
import com.smoothradio.radio.core.util.PlayerManager;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.android.qualifiers.ApplicationContext;
import dagger.hilt.components.SingletonComponent;

import javax.inject.Singleton;

@Module
@InstallIn(SingletonComponent.class)
public class AppModule {

    @Provides
    @Singleton
    public AppDatabase provideDatabase(@ApplicationContext Context context) {
        return Room.databaseBuilder(context, AppDatabase.class, "radio_db")
                .fallbackToDestructiveMigration()
                .build();
    }

    @Provides
    public RadioStationDao provideRadioStationDao(AppDatabase database) {
        return database.radioStationDao();
    }

    @Provides
    @Singleton
    public FirebaseFirestore provideFirebaseFirestore() {
        return FirebaseFirestore.getInstance();
    }
}

File: app/src/main/java/com/smoothradio/radio/core/di/AppModule.kt
package com.smoothradio.radio.core.di

import android.content.Context
import androidx.room.Room
import com.google.firebase.firestore.FirebaseFirestore
import com.smoothradio.radio.core.data.local.AppDatabase
import com.smoothradio.radio.core.data.local.RadioStationDao
import com.smoothradio.radio.core.data.repository.RadioRepository
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
class AppModule {

    @Provides
    @Singleton
    fun provideDatabase(@ApplicationContext context: Context): AppDatabase =
        Room.databaseBuilder(context, AppDatabase::class.java, "radio_db")
            .build()

    @Provides
    fun provideRadioStationDao(database: AppDatabase): RadioStationDao = database.radioStationDao()

    @Provides
    @Singleton
    fun provideFirebaseFirestore(): FirebaseFirestore = FirebaseFirestore.getInstance()

    @Provides
    @Singleton
    fun provideRadioRepository(dao: RadioStationDao): RadioRepository {
        return RadioRepository(dao)
    }
}


